### ISSUE TITLE
V11 support

### ISSUE THREAD

**Issue by pkonshik** [ironmonk108/enhanced-terrain-layer#142]
Hello, could you please take a look onto these errors on FVTT V11

![image](https://github.com/ironmonk88/enhanced-terrain-layer/assets/52590783/7dbedb81-8fe9-4b0f-b4c4-b33fe539c0a2)
![image](https://github.com/ironmonk88/enhanced-terrain-layer/assets/52590783/af7111f6-c52d-4a02-91d7-700e8c4e082f)

**Comment by airbreather** [comment]
I started playing around with this, and I managed to get through the first barrier by adding this wrapper in the `'init'` hook just before the `initApi()` call:
```javascript
let onSceneDefineSchema = function (wrapped, ...args) {
	return mergeObject(wrapped(...args), {
		Terrain: new foundry.data.fields.EmbeddedCollectionField(TerrainDocument),
	});
}

if (game.modules.get("lib-wrapper")?.active) {
	libWrapper.register("enhanced-terrain-layer", "Scene.defineSchema", onSceneDefineSchema, "WRAPPER");
} else {
	const oldDefineSchema = Scene.defineSchema;
	Scene.defineSchema = function (event) {
		return onSceneDefineSchema.call(this, oldDefineSchema.bind(this), ...arguments);
	}
}
```
which allowed me to change the `EmbeddedCollection` constructor call like so:
```diff
diff --git a/terrain-main.js b/terrain-main.js
index a274d23..bba96ba 100644
--- a/terrain-main.js
+++ b/terrain-main.js
@@ -324,7 +324,7 @@ async function addControlsv9(app, dest, { full = false, insert = false } = {}) {
 }
 
 function setupScene(scene) {
-       scene.terrain = new foundry.abstract.EmbeddedCollection(scene, [], Terrain);
+       scene.terrain = new foundry.abstract.EmbeddedCollection('Terrain', scene, []);
        let etl = scene.flags['enhanced-terrain-layer'];
        if (etl) {
                for (let [k, v] of Object.entries(etl)) {
```
From there, I had to make some changes in `TerrainShape` according to some signature changes in `DrawingShape` (mostly for the better), and that got some terrain objects from v10 to start showing up...

...but then I got kinda stuck where to go from there to make things actually start working more properly.  My guess is that my little hack to make `EmbeddedCollection` stop throwing errors isn't *quite* comprehensive enough to address the **reason** why the API changed so dramatically.

I'm going to come back to this later, but I figured I'd post this in case someone else who's more experienced with this Foundry stuff might have some insights?

**Comment by darkanx** [comment]
Anyone have any luck on this?

**Comment by rinnocenti** [comment]
I'm waiting a lot for this module to be able to migrate to v11

**Comment by Arthurneto1307** [comment]
> I'm waiting a lot for this module to be able to migrate to v11

me too :( i want this for v11

**Comment by Amd321** [comment]
Any word when this will be updated for version 11? In dire need of difficult terrain management :(

**Comment by ironmonk88** [comment]
Unfortunately... I can't get this to work.
v11 just ended up preventing too many things I need for it to work properly.

**Comment by LarusMascus** [comment]
Truly a shame... Note that in the meantime, you can always make do with an effect template to more or less make difficult terrain.
I'll hold on to the hope of being able to use it fully in the future!

**Comment by Rabulias** [comment]
>you can always make do with an effect template to more or less make difficult terrain.

Can you give more detail on how to do this?

**Comment by LarusMascus** [comment]
Sure, you can change the cost of movement in measured template. Just draw a Circle, Crone, Quare or Ray template and open the properties (double clic on the icon of the new template). You have a new category called Terrain in which you can change the cost of movement. It takes a lot of time depending on the size of the battlemap and the pattern of difficult terrain.
![image](https://github.com/ironmonk88/enhanced-terrain-layer/assets/141558125/984d2c77-fcaa-4043-85a1-05b8ab8a2782)

**Comment by EternalRider** [comment]
it has this problem.
![image](https://github.com/ironmonk88/enhanced-terrain-layer/assets/46736326/c57afc13-7531-4936-b4bd-2ea8e88da0a9)

**Comment by fishrobe** [comment]
> Unfortunately... I can't get this to work. v11 just ended up preventing too many things I need for it to work properly.

Aww that's a bummer. Didn't realize how much I used it until it was gone :(

Guess it's back to doing it in our heads until a better solution is found.

**Comment by LarusMascus** [comment]
True, using the rule directly give you the NaN error and you have to calculate the total yourself.
![image](https://github.com/ironmonk88/enhanced-terrain-layer/assets/141558125/373eba84-1622-4589-a7e0-d03ab212270d)

Nonetheless, using "Drag Ruler" mod (with "DnD5e Drag Ruler Integration" mod) it's working like a charm :)
![image](https://github.com/ironmonk88/enhanced-terrain-layer/assets/141558125/6e6e04f7-26ed-46e5-8adc-7c1ba75baf0d)

I personnaly put terrain ruler off normaly because i use the ruler to calculate distance for ranged attack or spell range/area (which doesn't care for difficult terrain).

**Comment by EternalRider** [comment]
能否将measured template的部分独立出来并修复这个总值计算问题？

**Comment by Rabulias** [comment]
You could use a special measurement token with a flying speed to measure without being impacted by terrain costs. Unless the difficult terrain cost extends above the ground...

**Comment by fishrobe** [comment]
> Sure, you can change the cost of movement in "measured template". Just draw a Circle, Cone, Square, or Ray template and open the properties (double click on the icon of the new template). You have a new category called Terrain in which you can change the cost of movement. It takes a lot of time depending on the size of the battlemap and the pattern of difficult terrain. ![image](https://user-images.githubusercontent.com/141558125/261872732-984d2c77-fcaa-4043-85a1-05b8ab8a2782.png)

I don't seem to get the terrain option here. Is it another module?
![measured template](https://github.com/ironmonk88/enhanced-terrain-layer/assets/50474856/dc508dd1-81f2-4d88-804f-7cb9b83c6195)

**Comment by Rabulias** [comment]
> I don't seem to get the terrain option here. (V11.308) Is it another module? ![measured template](https://user-images.githubusercontent.com/50474856/263204912-dc508dd1-81f2-4d88-804f-7cb9b83c6195.jpg)
Yes, you still need to have the Enhanced Terrain Layer module installed to get the Terrain tab on templates.
